iroha_crypto/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+use serde::{de::Error as SerdeError, Deserialize};
+use std::{
+    collections::BTreeMap,
+    convert::{TryFrom, TryInto},
+    fmt::{self, Debug, Display, Formatter},
+};
+use ursa::{
+    blake2::{
+        digest::{Input, VariableOutput},
+        VarBlake2b,
+    },
+    keys::{
+        KeyGenOption as UrsaKeyGenOption, PrivateKey as UrsaPrivateKey, PublicKey as UrsaPublicKey,
+    },
+    signatures::{ed25519::Ed25519Sha512, SignatureScheme, Signer},
+};
+
+pub const SIGNATURE_LENGTH: usize = 64;
+pub const HASH_LENGTH: usize = 32;
+pub const ED_25519: &str = "ed25519";
+
+/// Represents hash of Iroha entities like `Block` or `Transaction.
+pub type Hash = [u8; HASH_LENGTH];
+
+pub enum KeyGenOption {
+    UseSeed(Vec<u8>),
+    FromPrivateKey(PrivateKey),
+}
+
+impl TryFrom<KeyGenOption> for UrsaKeyGenOption {
+    type Error = String;
+
+    fn try_from(key_gen_option: KeyGenOption) -> Result<Self, Self::Error> {
+        match key_gen_option {
+            KeyGenOption::UseSeed(seed) => Ok(UrsaKeyGenOption::UseSeed(seed)),
+            KeyGenOption::FromPrivateKey(key) => {
+                if key.digest_function == ED_25519 {
+                    Ok(UrsaKeyGenOption::FromSecretKey(UrsaPrivateKey(key.payload)))
+                } else {
+                    Err(format!(
+                        "Ursa does not support {} digest function.",
+                        key.digest_function
+                    ))
+                }
+            }
+        }
+    }
+}
+
+/// Pair of Public and Private keys.
+#[derive(Clone, Debug, Deserialize, Default)]
+pub struct KeyPair {
+    /// Public Key.
+    pub public_key: PublicKey,
+    /// Private Key.
+    pub private_key: PrivateKey,
+}
+
+impl KeyPair {
+    /// Generates a pair of Public and Private key.
+    /// Returns `Err(String)` with error message if failed.
+    pub fn generate(key_gen_option: Option<KeyGenOption>) -> Result<Self, String> {
+        let (public_key, ursa_private_key) = Ed25519Sha512
+            .keypair(
+                key_gen_option
+                    .map(|key_gen_option| key_gen_option.try_into())
+                    .transpose()?,
+            )
+            .map_err(|e| format!("Failed to generate Ed25519Sha512 key pair: {}", e))?;
+        let public_key: [u8; 32] = public_key[..]
+            .try_into()
+            .map_err(|e| format!("Public key should be [u8;32]: {}", e))?;
+        let mut private_key = [0; 64];
+        private_key.copy_from_slice(ursa_private_key.as_ref());
+        Ok(KeyPair {
+            public_key: PublicKey {
+                digest_function: ED_25519.to_string(),
+                payload: public_key.to_vec(),
+            },
+            private_key: PrivateKey {
+                digest_function: ED_25519.to_string(),
+                payload: private_key.to_vec(),
+            },
+        })
+    }
+}
+
+/// Public Key used in signatures.
+#[derive(Encode, Decode, Ord, PartialEq, Eq, PartialOrd, Clone, Hash, Default)]
```

Comment:
It was great when `PublicKey` had a Copy trait. Can you make a small research and try to use `Sized` types for fields so we can Make PK Copy Again?)

Reply:
As we discussed on a meeting, we decided to store a String that would identify the algorithm for our internal representation. Therefore we had to use clone here... Do you have other suggestions?

Reply:
Because String is just an unsized array of valid UTF-8 bytes and we already had plans to replace it with something sized - what about defining some size which should be enough for encoding of this parameter?

Reply:
Ok, I will use u8 representation of an enum with specified values for now, similar to how it is done in multihash.